drivers/serial/uart_stm32.c

```diff
@@ -32,6 +32,387 @@
+		return UART_CFG_STOP_BITS_1_5;
+#endif	/* LL_USART_STOPBITS_1_5 */
+	case LL_USART_STOPBITS_2:
+	default:
+		return UART_CFG_STOP_BITS_2;
+	}
+}
+
+static inline u32_t uart_stm32_cfg2ll_databits(enum uart_config_data_bits db)
+{
+	switch (db) {
+/* Some MCU's don't support 7B datawidth */
+#ifdef LL_USART_DATAWIDTH_7B
+	case UART_CFG_DATA_BITS_7:
+		return LL_USART_DATAWIDTH_7B;
+#endif	/* LL_USART_DATAWIDTH_7B */
+	case UART_CFG_DATA_BITS_8:
+	default:
+		return LL_USART_DATAWIDTH_8B;
+	}
+}
+
+static inline enum uart_config_data_bits uart_stm32_ll2cfg_databits(u32_t db)
+{
+	switch (db) {
+/* Some MCU's don't support 7B datawidth */
+#ifdef LL_USART_DATAWIDTH_7B
+	case LL_USART_DATAWIDTH_7B:
+		return UART_CFG_DATA_BITS_7;
+#endif	/* LL_USART_DATAWIDTH_7B */
+	case LL_USART_DATAWIDTH_8B:
+	default:
+		return UART_CFG_DATA_BITS_8;
+	}
+}
+
+static int uart_stm32_configure(struct device *dev,
+				const struct uart_config *cfg)
+{
+	const u32_t parity = uart_stm32_cfg2ll_parity(cfg->parity);
+	const u32_t stopbits = uart_stm32_cfg2ll_stopbits(cfg->stop_bits);
+	const u32_t databits = uart_stm32_cfg2ll_databits(cfg->data_bits);
+	USART_TypeDef *UartInstance = UART_STRUCT(dev);
+
+	/* Hardware doesn't support mark or space parity */
+	if ((UART_CFG_PARITY_MARK == cfg->parity) ||
+	    (UART_CFG_PARITY_SPACE == cfg->parity)) {
+		return -ENOTSUP;
+	}
+
+#if defined(LL_USART_STOPBITS_0_5) && defined(CONFIG_LPUART_1)
+	if (IS_LPUART_INSTANCE(UartInstance) &&
+	    UART_CFG_STOP_BITS_0_5 == cfg->stop_bits) {
+		return -ENOTSUP;
+	}
+#else
+	if (UART_CFG_STOP_BITS_0_5 == cfg->stop_bits) {
+		return -ENOTSUP;
+	}
+#endif
+
+#if defined(LL_USART_STOPBITS_1_5) && defined(CONFIG_LPUART_1)
+	if (IS_LPUART_INSTANCE(UartInstance) &&
+	    UART_CFG_STOP_BITS_1_5 == cfg->stop_bits) {
+		return -ENOTSUP;
+	}
+#else
+	if (UART_CFG_STOP_BITS_1_5 == cfg->stop_bits) {
+		return -ENOTSUP;
+	}
+#endif
+
+	/* Driver doesn't support 5 or 6 databits and potentially 7 */
+	if ((UART_CFG_DATA_BITS_5 == cfg->data_bits) ||
+	    (UART_CFG_DATA_BITS_6 == cfg->data_bits)
+#ifndef LL_USART_DATAWIDTH_7B
+	    || (UART_CFG_DATA_BITS_7 == cfg->data_bits)
+#endif /* LL_USART_DATAWIDTH_7B */
+		) {
+		return -ENOTSUP;
+	}
+
+	/* Driver currently doesn't support line ctrl */
+	if (UART_CFG_FLOW_CTRL_NONE != cfg->flow_ctrl) {
+		return -ENOTSUP;
+	}
+
+	LL_USART_Disable(UartInstance);
+
+	if (parity != uart_stm32_get_parity(dev)) {
+		uart_stm32_set_parity(dev, parity);
+	}
+
+	if (stopbits != uart_stm32_get_stopbits(dev)) {
+		uart_stm32_set_stopbits(dev, stopbits);
+	}
+
+	if (databits != uart_stm32_get_databits(dev)) {
+		uart_stm32_set_databits(dev, databits);
+	}
+
+	if (cfg->baudrate != uart_stm32_get_baudrate(dev)) {
+		uart_stm32_set_baudrate(dev, cfg->baudrate);
+	}
+
+	LL_USART_Enable(UartInstance);
+	return 0;
+};
+
+static int uart_stm32_config_get(struct device *dev, struct uart_config *cfg)
+{
+	cfg->baudrate = uart_stm32_get_baudrate(dev);
```

Comment:
I had a discussion with @erwango long days ago, here is the link: #8306
(don't know why can't copy the link now, so I copied them)

> 

    Not sure we need to use data fields to store information, since we can use device registers.

No, you should not do that. We must using an variable to store the baud_rate setted before.
the baud_rate we read using LL_USART_GetBaudRate maybe not equal the baud_rate setted using LL_USART_SetBaudRate because the BRR is inaccurate for some specify baud rate.

eg: I had set uart's baud rate to 115200, but the baudrate readed is 115107.

332             baud_rate = LL_USART_GetBaudRate(UartInstance, clock_rate, LL_USART_OVERSAMPLING_16);

(gdb) p baud_rate 
$1 = 115107

So in my options, we should use a variable to save the baud rate.

Reply:
Right, my comment was for other fields (parity and so on), I guess you agree for those ?

Reply:
yes😊

Reply:
I didn't realize that could occur. I've pushed new code, please recheck.